src/loki_cluster.py

```diff
@@ -0,0 +1,317 @@
+"""
+
+import json
+import logging
+from enum import Enum, unique
+from typing import Any, Dict, Iterable, List, MutableMapping, Optional, Set
+from urllib.parse import urlparse
+
+import ops
+import pydantic
+from ops import EventSource, Object, ObjectEvents, RelationCreatedEvent
+from pydantic import BaseModel, ConfigDict
+
+log = logging.getLogger("loki_cluster")
+
+DEFAULT_ENDPOINT_NAME = "loki-cluster"
+BUILTIN_JUJU_KEYS = {"ingress-address", "private-address", "egress-subnets"}
+
+
+# TODO: inherit enum.StrEnum when jammy is no longer supported.
+# https://docs.python.org/3/library/enum.html#enum.StrEnum
+@unique
+class LokiRole(str, Enum):
+    """Loki component role names."""
+
+    # Meta roles based on:
+    # https://grafana.com/docs/loki/latest/get-started/deployment-modes/#simple-scalable
+    read = "read"
+    write = "write"
+    backend = "backend"
+    all = "all"
+
+
+META_ROLES = {}
+
+
+def expand_roles(roles: Iterable[LokiRole]) -> Set[LokiRole]:
+    """Expand any meta roles to their 'atomic' equivalents."""
+    expanded_roles = set()
+    for role in roles:
+        if role in META_ROLES:
+            expanded_roles.update(META_ROLES[role])
+        else:
+            expanded_roles.add(role)
+    return expanded_roles
+
+
+class ConfigReceivedEvent(ops.EventBase):
+    """Event emitted when the "loki-cluster" provider has shared a new loki config."""
+
+    config: Dict[str, Any]
+    """The loki config."""
+
+    def __init__(self, handle: ops.framework.Handle, config: Dict[str, Any]):
+        super().__init__(handle)
+        self.config = config
+
+    def snapshot(self) -> Dict[str, Any]:
+        """Used by the framework to serialize the event to disk.
+
+        Not meant to be called by charm code.
+        """
+        return {"config": json.dumps(self.config)}
+
+    def restore(self, snapshot: Dict[str, Any]):
+        """Used by the framework to deserialize the event from disk.
+
+        Not meant to be called by charm code.
+        """
+        self.relation = json.loads(snapshot["config"])
+
+
+class LokiClusterError(Exception):
+    """Base class for exceptions raised by this module."""
+
+
+class DataValidationError(LokiClusterError):
+    """Raised when relation databag validation fails."""
+
+
+class DatabagAccessPermissionError(LokiClusterError):
+    """Raised when a follower attempts to write leader settings."""
+
+
+class JujuTopology(pydantic.BaseModel):
+    """JujuTopology."""
+
+    model: str
+    unit: str
+    # ...
+
+
+class DatabagModel(BaseModel):
+    """Base databag model."""
+
+    model_config = ConfigDict(
+        # Allow instantiating this class by field name (instead of forcing alias).
+        populate_by_name=True,
+        # Custom config key: whether to nest the whole datastructure (as json)
+        # under a field or spread it out at the toplevel.
+        _NEST_UNDER=None,
```

Comment:
This is a new charm. Should we get rid of the _NEST_UNDER logic? Seems like a retrofit we don't need here.
One option is to have the lib's api version ("v0") as the default "nest under".

Reply:
Yes, I prefer to have the whole datastructure under a field instead of spread it out at the toplevel.
Then, let's use nest_under as default